Jorislive patch optional region

jorislive opened the pull request:
When your database was created in the region 'eu' or 'us' specificly the setup won't run, so I've added an optional domain parameter for the setup client.

jorislive committed:
Update setup.js

When your database was created in the region 'eu' or 'us' specificly the setup won't run, so I've added an optional domain parameter for the setup client

jorislive committed:
Update .env.local.example

rts-rob commented on scripts/setup.js:
```diff
@@ -52,7 +54,7 @@ const main = async () => {
     console.log(explanation)
   }
-  let client = new faunadb.Client({ secret: adminKey })
+  let client = new faunadb.Client({ secret: adminKey, scheme: 'https', domain })
```
Can you pull out the hardcoded 'scheme' please?

jorislive replied:
👍

rts-rob commented on .env.local.example:
```diff
@@ -1,4 +1,5 @@
 REACT_APP_LOCAL___ADMIN=
+REACT_APP_LOCAL___ADMIN_DOMAIN=<optional 'db.eu.fauna.com' or 'db.us.fauna.com'>
```
Please move these optional values to a comment around line 37 in scripts/setup.js 

This way, anyone using the classic Region Group doesn't need to set a value here.

Line should read:
```
REACT_APP_LOCAL___ADMIN_DOMAIN=
```

jorislive replied:
👍

rts-rob reviewed (changes requested):
Hey @jorislive - thank you for submitting this!

I've requested a few convention-driven changes. With those in place, we'll be g2g.

jorislive committed:
fix: follow up on comment rts-rob

jorislive committed:
fix: auth add missing import statement

jorislive committed:
feature: add domain functionality also to query-manager and rename environment key

jorislive commented:
Hi Rob,

I followed up on your comments and also added the domain functionality to the query manager. Furthermore I came accross a missing import statement and fixed that as well.

Cheers Joris

rts-rob commented on src/fauna/queries/auth.js:
```diff
@@ -1,4 +1,4 @@
-import faunadb from 'faunadb'
+import faunadb, {Identify} from 'faunadb'
```
Good catch; thanks!

rts-rob reviewed (changes requested):
Thanks for catching the missing import.

This PR needs modifications to src/fauna/queries/*.spec.js (auth, fweets, rate-limiting, search) to make all the tests pass with US and EU keys.

If you're willing to make those updates, that would be great. If that's too big an ask, I'll put them in. Simplest way would be to add me as a contributor to this fork, if you're okay with that. Otherwise I'll fork it and PR back into here.

jorislive committed:
feature: add domain functionality also to ./src/fauna/queries/*.spec.js queries

jorislive commented:
> Thanks for catching the missing import.
> 
> This PR needs modifications to src/fauna/queries/*.spec.js (auth, fweets, rate-limiting, search) to make all the tests pass with US and EU keys.

I've implemented the domain functionality (see last commit), but I wasn't able to test it properly, I've tried to run tests (both in a git clone with 'regional' database and a git clone with 'classic' database. Since there's no instruction about setting up the test-suite I assumed that I needed to create a .env.test.local file with the same secret as used in the .env.local file entry REACT_APP_LOCAL___ADMIN, but I wasn't able to pass all the tests. So I'm doing something wrong..

rts-rob committed:
Add support for region group endpoints to test suite.

WIP: At this point all tests pass *eventually*, but US end EU tests are flaky. Do not merge until resolved.

rts-rob committed:
Merge commit: Still WIP: DO NOT MERGE PR

rts-rob committed:
WIP: DO NOT MERGE: Removed duplicate domain definitions.

rts-rob commented:
Thanks, @jorislive - I think the difference was using three underscores for the domain vs. two as defined in .env.test.local.example

I've pushed a few changes. These all work for certain definitions of "work," but the US and EU tests are intermittently failing. My hypothesis is that this is a test setup issue, where we're creating dependent resources in parallel, and the smaller Region Groups are just too fast. Great problem to have!

I'm going to have a look at serializing the resource setup. If that works, we should be g2g.

rts-rob committed:
WIP: Factoring out test db connection to work on contention in Region Groups

jorislive commented:
@rts-rob 

> Thanks, @jorislive - I think the difference was using three underscores for the domain vs. two as defined in .env.test.local.example
> 
> I've pushed a few changes. These all work for certain definitions of "work," but the US and EU tests are intermittently failing. My hypothesis is that this is a test setup issue, where we're creating dependent resources in parallel, and the smaller Region Groups are just too fast. Great problem to have!
> 
> I'm going to have a look at serializing the resource setup. If that works, we should be g2g.

I'm not really sure if I fully understand that 'too fast' could be problematic, but it does sound like a luxury problem indeed!
Good luck with serializing the resource setup! 

It was nice for me to have a look at the code, it provided me with good insights in how to create a more advanced setup with Fauna. For me Fauna was already the weapon of choice for the client of my new project, so I needed to deep dive into it. But I must say I'm starting to become a real fan of it's abillities!

rts-rob committed:
WIP: Factored out createDatabase and deleteDatabase

rts-rob commented:
Hello again @jorislive - we're going to close this PR in favor of #79 (just merged).

79 takes the approach of recommending users run this in the Classic/Global Region Group. This is a simple documentation fix, and more appropriate for a learning project.

We explored making the fixes to enable Fwitter to run in all Region Groups, but it requires a significant refactor to get the tests to pass. This refactor would be moving schema generation onto [Fauna Schema Migrate](https://github.com/fauna-labs/fauna-schema-migrate) to avoid resource creation contention issues in the smaller, faster Region Groups.

We're currently working on a new, opinionated flagship application that incorporates Region Group support from the beginning. This is expected to be released by end of October 2021, with additional features layered in throughout the coming year. The level of effort required to move Fwitter to FSM is too high when that work will ultimately be replaced in less than a quarter.

Thanks again for submitting this PR!
